README.md
**Toggler** manages App state booleans: with easy state transitions, pre-commit state validation, and post-commit change notifications.

**Toggler** object keeps state of up to 52 boolean values (items) that can be manipulated one by one or in concert. _Radio group_ behaviour can be declared on up to 17 separated groups of items. Independent _disabled_ property is avaliable for every item – to be used from within UI builders.

**Toggler** was designed specifically for singleton _Models_ and for _observer_ style state flows, though it can also be used in _reactive_ state management architectures via its `state` and `clone` copying constructors.  For safe use within a singleton _Model_ Toggler has built-in data race detection and automatically skips changes coming from an outdated ancestor state.

Toggler is a single class library with no dependencies.

Test coverage: `100.0% (145 of 145 lines)`

## Getting started

 1. `$> dart pub get toggler`
 1. import toggler for your Model: `import 'package:toggler/toggler.dart';`
 2. declare meaningful names for your knobs:
 ```Dart
     const tgTurn = 0; // min Toggler item index
     const tgOther = 1; // ...
     const tgClaim = tgIndexMax; // max Toggler item index (51)
 ```
 4. add a ValueNotifier `final fchg = ValueNotifier<int>(0)` to your Model
 4. add Toggler `final tog = Toggler(notify: (Toggler _, Toggler n) => fchg.value = n.serial);`
 4. wire it to your UI code (_example with [get_it_mixin](https://pub.dev/packages/get_it_mixin)_):

```Dart
  // ...somewhere in your Widget tree:
  Widget build(BuildContext context) {
    final tog = getX((Model m) => m.tog);
    watchX((Model m) => m.fchg);
    // ...
    return tog[tgTurn] // somewhere in build
      ? const IconYou(...)
      : const IconOpponent(...),
            // ... ClaimPrize is shown disabled unless active
            onPressed: tog.active(tgClaim)
              ? () => tog.toggle(tgClaim)
              : null,
```

### State flow

1. somewhere in your App code a state of a single item in a _live_ Toggler in Model is changed by a `toggle(flagName)` call (or other _state setters_: `set`, `clear`, `enable`, `disable`).
2. State transition function `bool fix(oldState, newState)` is called next,
3. then "fixed" new state is commited to the state of Toggler instance,
4. then `notify(oldState, liveState)` is called, there Model's _ValueNotifiers_ can be updated or _notifyListeners_ called to pass "changed" baton to the Presentation layer.

If `fix` handler has not been provided, any single item change made by setter is commited immediately. Otherwise `fix` is called with _newState_ reflecting change that came from a setter. Then `fix` returns _true_, to have _newState_ commited; Or _false_ for changes to be abandoned.

Your code in `fix` may manipulate _newState_ at will, it even may assign a some predefined const values to the `tg` and/or `ds` properties of it. Usually `fix` is Model's internal function, so it may have access to all other pieces of your business logic.

After succesful new state commit `notify` is called. Within this handler you may selectively check where changes were made, eg using `recent` index, or `differsFrom(oldState, rangeFirst, rangeLast)` helper. Then you may run your chosen state passing machinery (Eg. update InheritedWidget state, feed an EventObserver, push a cloned object to the Stream...).
> Toggler author's preffered mechanics: copy `serial` of a new state to the [ValueNotifier](https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html).value that is observed in the **stateless** widgets tree using [get_it_mixin](https://pub.dev/packages/get_it_mixin)).


### API 101

#### constructors:
- `Toggler({fix: onChange, notify: afterChange, tg: 0, ds: 0, rm: 0, cm: 0, hh: 0})`
  > at least `notify` is needed to make a _live_ Toggler. All other members can be given to default constructor, too - used eg. in saved state deserializer and tests. An all-default Toggler can be mutated at will, eg. in an explicit App state initializer, then `notify` and/or `fix` handlers can be attached later.
- `state()` method returns a _copy of state_ only (ie. with `fix` = `notify` = null).
- `clone()` method returns a deep copy of `this`. _Caveat emptor!_

#### getters:
- `[i]` returns state of item at index i (_true_ for a set item).
- `active(i)` returns negation of _disabled_ property of item at index i.

#### setters:
- `toggle(i)`, `set(i)`, `clear(i)`, `setTo(i, state)`
  > mutate a single item state at index i.
- `toggle(i, ifActive: true)`, `set(i, ifActive: true)`, `...`
  > item state setters may depend on _disabled_ property.
- `enable(i)`, `disable(i)`, `setDS(i, state)`
  > mutate _disabled_ property of item at index i.

#### state tests:
- `recent` is index of a latest singular change.
  > In a `fix` function this means change that fired it.
- `serial` is a monotonic state serial number (34b), bigger is newer
- `isOlderThan(other)` compares serial numbers of state copies.
  > Any _live_ object always is newer than any other Toggler object.
- `changed(i, relmask)`
  > returns _true_ if there was a recent change at _i_ or _relmask_ matching positions
- `anyInSet({first, last, relmask})`
  > returns _true_ if any value in _first..last_ range, or on _relmask_ matching positions is _set_.
- `differsFrom(other, {first, last, relmask})`
  > compares both value and _disabled_ property of _this_ and _other_ item, returns _true_ if any bit differs. Check can be limited to _first..last_ index range, or to _relmask_ matching positions.
- `cm` change mask has bit set to 1 at position(s) of latest change(s)

#### diagnostics:
- `error`
  > set to _true_ if any Toggler method taking `i` index number got it out of valid range 0..max.
- `race`
  > set to _true_ if `fix` of older generation state unsuccesfully tried to update a _live_ object, eg. after `fix` being suspended by await.
- `done`
  > can be set by your App code to signal that changes were seen/used. This flag is automatically cleared on every new change.

#### radio group setup:
- `radioGroup(first, last)`
  > makes items in _first..last_ range be dependent on each other, so if one is set any other in group will be cleared automatically. Up to 17 separate groups can be defined - a single group per each `radioGroup` call. RadioGroup ranges may not overlap nor be adjacent.

#### serialize:
- Toggler intentionally has no `toString` nor `toJSON` methods. Its whole state is public and consist of just four _ints_. It should be handled seamlesslly by any serialization method one may have chosen for a whole _Model_.

#### state serial overflow:
- Serial counter will overflow once per 16 billion changes. It is 500 years of once per second user taps, 9 years of per-frame toggles @60fps, or 6 months of counting every network frame @12Mbps. If overflow happens, the `isOlderThan` method may lie comparing state copies made before and after overflow. (_IOW: it may concern you if you use _reactive_ style state management __and__ use serial as identity-age sentinel __and__ your code is meant to work 24/365. Then you may consider resetting `hh` to 0 at suitable moments eg. when your input queque empties_).


CHANGELOG.md
## 1.0.0

- Initial public version of Toggler library. With CLI example.
  Test coverage: 100%, covered: 133 code lines (of 204 code lines total)


test/toggler_test.dart
// Copyright (c) 2022 Wojciech S. Czarnecki, aka Ohir Ripe

import 'package:toggler/toggler.dart';
import 'package:test/test.dart';

final Matcher throwsAssertionError = throwsA(isA<AssertionError>());
void main() {
  group('Rudimentary :: ', () {
    final flags = Toggler();

    setUp(() {
      flags.tg = flags.ds = flags.rg = flags.hh = flags.chb = 0; // reset
      flags.fix = flags.notify = null;
    });

    test('Set 0 Max', () {
      flags[0] = true; // cover []= setter
      flags.set(tgIndexMax);
      expect(
          flags[0] && !flags[1] && !flags[tgIndexMax - 1] && flags[tgIndexMax],
          isTrue);
      flags.clear(0);
      flags.clear(tgIndexMax);
      expect(!flags[0] && !flags[tgIndexMax], isTrue);
    });
    test('Is Set in range set boundry', () {
      flags.set(7);
      flags.set(12);
      expect(flags.anyInSet(first: 7, last: 10), isTrue);
      expect(flags.anyInSet(first: 8, last: 12), isTrue);
    });
    test('Is Set in range not set boundary', () {
      flags.set(7);
      flags.set(12);
      expect(flags.anyInSet(first: 8, last: 11), isFalse);
    });
    test('DisableEnable', () {
      flags.disable(0);
      expect(flags.active(0), isFalse);
      flags.set(0, ifActive: true);
      expect(flags[0], isFalse);
      flags.enable(0);
      flags.set(0, ifActive: true);
      expect(flags[0], isTrue);
    });
    test('setDisable setEnable', () {
      flags.setDS(0, true);
      expect(flags.active(0), isFalse);
      flags.set(0, ifActive: true);
      expect(flags[0], isFalse);
      flags.setDS(0, false);
      flags.set(0, ifActive: true);
      expect(flags[0], isTrue);
    });
    test('Disable Max', () {
      flags.set(tgIndexMax);
      flags.disable(tgIndexMax);
      flags.clear(tgIndexMax, ifActive: true);
      flags.setTo(tgIndexMax, false, ifActive: true);
      expect(flags[tgIndexMax], isTrue);
      flags.setTo(tgIndexMax, false);
      expect(flags[tgIndexMax], isFalse);
    });
    test('DisableEnable Max-1', () {
      flags.disable(tgIndexMax - 1);
      flags.set(tgIndexMax - 1, ifActive: true);
      expect(flags[tgIndexMax - 1], isFalse);
      flags.setTo(tgIndexMax - 1, true, ifActive: true);
      expect(flags[tgIndexMax - 1], isFalse);
      flags.enable(tgIndexMax - 1);
      flags.set(tgIndexMax - 1, ifActive: true);
      expect(flags[tgIndexMax - 1], isTrue);
    });
    test('DisableEnable 33', () {
      flags.set(33, ifActive: true);
      flags.disable(33);
      flags.clear(33, ifActive: true);
      expect(flags[33], isTrue);
      flags.setTo(33, false, ifActive: true);
      expect(flags[33], isTrue);
      flags.enable(33);
      flags.clear(33, ifActive: true);
      expect(flags[33], isFalse);
    });
    test('Toggle 0', () {
      flags.disable(0);
      flags.toggle(0, ifActive: true);
      expect(flags[0], isFalse);
      flags.toggle(0);
      expect(flags[0], isTrue);
      flags.toggle(0);
      expect(flags[0], isFalse);
    });
    // test('name ', () {});
    test('generational behaviour [A]', () {
      var cLive = Toggler(notify: (Toggler a, Toggler b) => a.isOlderThan(b));
      var c0 = cLive.state();
      var c1 = cLive.state();
      cLive.set(7);
      var c2 = cLive.state();
      cLive.clear(1);
      expect(c0.isOlderThan(c1), isFalse); // copy to copy
      expect(c1.isOlderThan(c2), isTrue); // copy to copy
      expect(c2.isOlderThan(c1), isFalse); // copy to copy
      expect(c2.isOlderThan(c2), isFalse); // copy to copy
    });
    test('generational behaviour [B]', () {
      var cLive = Toggler(notify: (Toggler a, Toggler b) {});
      var c0 = cLive.state();
      var c1 = cLive.state();
      cLive.set(7);
      var cL2 = Toggler(notify: (Toggler a, Toggler b) {});
      var c4 = cLive.state();
      expect(cLive.isOlderThan(cL2), isFalse); // live to live, always false
      expect(c4.isOlderThan(cL2), isTrue); // copy to live, always true

      c0.set(15);
      c1.set(33);
      expect(c0.hh == c1.hh, isTrue); // copies may not alter history
      expect(c0.serial == c1.serial, isTrue);
      expect(c0.recent == c1.recent, isTrue);
    });
    test('copy may not mutate history', () {
      var c0 = Toggler();
      c0.disable(3);
      c0.set(2);
      var c1 = c0.state();
      c1.set(16);
      var c2 = c1.state();
      expect(
          c0[2] &&
              c1[2] &&
              c2[2] &&
              c1[16] &&
              c2[16] &&
              c0.hh == c1.hh &&
              c2.hh == c0.hh &&
              c2.hh == 0,
          isTrue);
    });
    test('Change mask A', () {
      flags.set(0);
      expect(flags.chb == 1, isTrue);
      flags.set(10);
      expect(flags.chb == 1024, isTrue);
      expect(flags.changed(i: 10), isTrue);
      expect(flags.changed(i: 10, relmask: 7 << 8), isTrue);
      expect(flags.changed(i: 11, relmask: 7 << 8), isFalse);
      expect(flags.changed(i: 10, relmask: 7 << 7), isFalse);
      expect(flags.changed(relmask: 7 << 8), isTrue);
      flags.set(11);
      flags.set(33);
      flags.disable(51); // cm must reflect any change at index
      expect(flags.chb == 1 << 51, isTrue);
    });
    test('Change mask B', () {
      bool cf(Toggler o, Toggler n) {
        n.set(1);
        n.disable(2);
        return true;
      }

      flags.fix = cf;
      flags.set(0);
      expect(flags.chb == 7, isTrue); // b0,b1,b2 changed
    });
    test('Differs', () {
      var c1 = flags.state();
      expect(c1.differsFrom(flags), isFalse);
      c1.set(5);
      expect(c1.differsFrom(flags), isTrue);
      expect(c1.differsFrom(flags, relmask: 63), isTrue);
      expect(c1.differsFrom(flags, relmask: 31), isFalse);
      flags.set(5);
      expect(c1.differsFrom(flags), isFalse);
      flags.set(8);
      c1.set(23);
      expect(c1.differsFrom(flags, first: 8, last: 22), isTrue);
      expect(c1.differsFrom(flags, first: 9, last: 22), isFalse);
      expect(c1.differsFrom(flags, first: 9, last: 23), isTrue);
    });
    test('Set 63 | should throw', () {
      expect(() {
        flags.set(63);
      }, throwsAssertionError);
    });
    test('Set -1 | should throw', () {
      expect(() {
        flags.set(-1);
      }, throwsAssertionError);
    });
  });
  group('Radio :: ', () {
    final flags = Toggler();

    setUp(() {
      flags.tg = flags.ds = flags.rg = flags.hh = 0; // reset
      flags.radioGroup(kA, kC);
      flags.radioGroup(kE, kH);
      flags.set(kA);
      flags.set(kH);
    });
    test('Set overlapping radio | should throw', () {
      expect(() {
        flags.radioGroup(kI, kL);
      }, throwsAssertionError);
    });
    test('Set single radio | should throw', () {
      expect(() {
        flags.radioGroup(kL, kL);
      }, throwsAssertionError);
    });
    test('Set 0..3 radio', () {
      flags.rg = 0; // reset radios
      flags.radioGroup(k0, kC);
      flags.set(k0);
      expect(flags[k0], isTrue);
      flags.set(kC);
      expect(!flags[k0] && !flags[kA] && !flags[kB] && flags[kC], isTrue);
      flags.set(kB);
      expect(!flags[k0] && !flags[kA] && flags[kB] && !flags[kC], isTrue);
      flags.set(kA);
      expect(!flags[k0] && flags[kA] && !flags[kB] && !flags[kC], isTrue);
    });
    test('Set 60..63 radio | should throw', () {
      expect(() {
        flags.rg = 0; // reset radios
        flags.radioGroup(60, 63);
      }, throwsAssertionError);
    });
    test('A H should be set in setup', () {
      expect(flags[kA] && flags[kH] && !flags[k0] && !flags[kI], isTrue);
    });
    test('set C E', () {
      flags.set(kC);
      flags.set(kE);
      expect(flags[kC] && flags[kE] && !flags[kA] && !flags[kH], isTrue);
    });
    test('set B F', () {
      flags.set(kB);
      flags.set(kF);
      expect(flags[kB] && flags[kF] && !flags[kA] && !flags[kH], isTrue);
    });
    test('set D I', () {
      flags.set(kD);
      flags.set(kI);
      expect(flags[kA] && flags[kH] && flags[kD] && flags[kI], isTrue);
    });
  });
  group('Diagnostics :: ', () {
    final flags = Toggler();

    setUp(() {
      flags.tg = flags.ds = flags.rg = flags.hh = 0; // reset
    });
    test('Race/Err/Done set true', () {
      flags.error = true;
      flags.race = true;
      flags.done = true;
      expect(flags.error, isTrue);
      expect(flags.race, isTrue);
      expect(flags.done, isTrue);
    });
    test('Race/Err/Done set false', () {
      flags.error = true;
      flags.race = true;
      flags.done = true;
      flags.error = false;
      flags.race = false;
      flags.done = false;
      expect(flags.error, isFalse);
      expect(flags.race, isFalse);
      expect(flags.done, isFalse);
      flags.setDone();
      expect(flags.done, isTrue);
    });
    test('Demand bad diff| should throw', () {
      expect(() {
        flags.differsFrom(flags.state(), last: 63);
      }, throwsAssertionError);
    });
  });
  group('Done and history :: ', () {
    // ensure
    int ntlast = 0;
    int filast = 0;
    void chnote(Toggler oS, Toggler nS) => ntlast++;

    bool chfix(Toggler oS, Toggler nS) {
      filast++;
      return true;
    }

    final flags = Toggler();
    const ohh = 7777777777;
    setUp(() {
      flags.tg = flags.ds = flags.rg = 0;
      flags.hh = ohh;
      flags.fix = chfix;
      flags.notify = chnote;
      ntlast = 0;
      filast = 0;
    });

    test('no history changes', () {
      flags.notify = null;
      flags.fix = null;
      flags.toggle(5);
      expect(flags[5] && flags.hh == ohh, isTrue);
    });
    test('done is cleared with copy', () {
      flags.setDone();
      final c1 = flags.state();
      expect(flags != c1, isTrue);
      expect(flags.done, isTrue);
      expect(c1.done, isFalse);
    });
    test('done is cleared with clone', () {
      flags.setDone();
      final c1 = flags.clone();
      expect(flags.done, isTrue);
      expect(c1.done, isFalse);
    });
    test('done is cleared on copy of copy', () {
      final c1 = flags.state();
      c1.done = true;
      expect(c1.done, isTrue);
      final c2 = c1.state();
      expect(c2.done, isFalse);
    });
    test('history changes with just fix', () {
      flags.notify = null;
      flags.toggle(5);
      expect(flags[5], isTrue);
      expect(flags.hh != ohh, isTrue);
      expect(filast, equals(1));
    });
    test('history changes with just notify', () {
      flags.fix = null;
      flags.toggle(5);
      expect(flags[5] && flags.hh != ohh, isTrue);
      expect(ntlast, equals(1));
    });
    test('done is cleared on fix only', () {
      flags.notify = null;
      expect(flags.done, isFalse);
      flags.setDone();
      expect(flags.done, isTrue);
      flags.toggle(5);
      expect(flags.done, isFalse);
      flags.done = true;
      expect(flags.done, isTrue);
      flags.toggle(5);
      expect(flags.done, isFalse);
      expect(!flags[5] && flags.hh != ohh, isTrue);
      expect(ntlast, equals(0));
      expect(filast, equals(2));
    });
    test('done is cleared on notify only', () {
      flags.fix = null;
      expect(flags.done, isFalse);
      flags.setDone();
      expect(flags.done, isTrue);
      flags.toggle(5);
      expect(flags.done, isFalse);
      flags.done = true;
      expect(flags.done, isTrue);
      flags.toggle(5);
      expect(flags.done, isFalse);
      expect(!flags[5] && flags.hh != ohh, isTrue);
      expect(ntlast, equals(2));
      expect(filast, equals(0));
    });
    /*
    */
  });

  group('FixNotifyRaces :: ', () {
    int last = 0;
    void chnote(Toggler oS, Toggler nS) => last++;

    bool chfix(Toggler oS, Toggler nS) {
      if (oS.recent == 25) {
        oS.hh <<= 1; // test abandon older state
      }
      oS.differsFrom(nS, first: 11, last: 16); // cover !differs path
      if (nS[1] && nS.differsFrom(oS)) nS.set(0); // test state fixing on 1
      if (nS[7] && nS.differsFrom(oS)) nS.setDone(); // test skip notify on 7
      if (nS[9] && nS.differsFrom(oS)) nS.done = true; // test skip notify on 9
      return true;
    }

    final flags = Toggler(notify: chnote, fix: chfix);

    setUp(() {
      flags.tg = flags.ds = flags.rg = flags.hh = 0; // reset
    });
    test('Notify 0', () {
      flags.set(0);
      expect(flags[0] && !flags[1], isTrue);
      expect(last == 1, isTrue);
    });
    test('Notify 1', () {
      flags.set(1);
      expect(last == 2, isTrue);
      expect(flags[0] && flags[1], isTrue);
    });
    test('Notify via clone', () {
      var nf = flags.clone();
      nf.set(5);
      expect(nf[5] && !flags[5], isTrue);
      expect(last == 3, isTrue);
    });
    test('Internal fix', () {
      flags.set(7);
      expect(last == 3, isTrue); // notify skipped, setDone()
      flags.set(9);
      expect(last == 3, isTrue); // notify skipped, done = true
    });
    test('Make artificial race | should throw', () {
      expect(() {
        flags.set(25);
        flags.set(0);
      }, throwsAssertionError);
    });
    // TODOx Make real racing test to hit `if (hh != newS.hh)` in _ckFix
  });
}

const k0 = 0;
const kA = 1;
const kB = 2;
const kC = 3;
const kD = 4;
const kE = 5;
const kF = 6;
const kG = 7;
const kH = 8;
const kI = 9;
const kJ = 10;
const kK = 11;
const kL = 12;
const kM = 13;
const kN = 14;
const kO = 15;
const kP = 16;
const kQ = 17;
const kR = 18;
const kS = 19;
const kT = 20;
const kU = 21;
const kW = 22;
const kV = 23;
const kX = 24;
const kY = 25;
const kZ = 26;


example/togglers_example.dart
import 'package:togglers/togglers.dart';

void main() {
  void ourNotify(_, Togglers cu) {
    var tg = '    set: ';
    var ds = ' active: ';
    for (int i = 0; i < 27; i++) {
      tg += cu[i] ? ' ^' : ' -';
      ds += cu.hasActive(i) ? ' .' : ' !';
    }
    print('kTGname:  0 A B C D E F G H I J K L M N O P Q R S T U W V X Y Z');
    print(tg);
    print(ds);
    print('          -----------------------------------------------------');
  }

  /*void ourChgNotifier(Togglers oS, Togglers nS) {
    // final chVN = ValueNotifier<int>(0);
    if (toggledInRange(oS, nS, first: kTGnameC, last: kTGnameH)) {
      // chVN = nS.hh // hh is guaranteed to be unique on each change
    }
  }*/

  // pt is a copy of previous state, cu is the current (live) state
  bool ourCheck(Togglers oS, Togglers nS) {
    // validate: 'NameA' can toggle only if 'Name0' was set.
    if (!oS[kTGname0] && oS[kTGnameA] != nS[kTGnameA]) {
      print('      >>> NameA change supressed by validator');
      return false; // disallow change to 'NameA'
    }
    // fix B,C if F radio was toggled
    if (oS[kTGnameF] != nS[kTGnameF]) {
      if (nS[kTGnameF]) {
        nS.disable(kTGnameB);
        nS.disable(kTGnameC);
        nS.clear(kTGnameB);
        nS.clear(kTGnameC);
      } else {
        nS.enable(kTGnameB);
        nS.enable(kTGnameC);
      }
    }
    return true; // accept changes
  }

  final flags = Togglers(notify: ourNotify, checkFix: ourCheck);

  // declare a radioGroup, up to 21 groups can be made over 63 items.
  flags.radioGroup(kTGnameD, kTGnameF);
  // fiddle:
  print('Trying to set A (ourCheck validator disallows this)');
  flags.set(kTGnameA);
  print('Set 0, NameA can be set only if Name0 was set before');
  flags.set(kTGname0);
  print('Now A is allowed to be toggled');
  flags.set(kTGnameA);
  print('Set B');
  flags.set(kTGnameB);
  print('Set C');
  flags.set(kTGnameC);
  print('Set D (of D..F radio group)');
  flags.set(kTGnameD);
  print('Set E of radio D..F - D will clear automatically.');
  flags.set(kTGnameE);
  print('Set F radio. E will clear then B and C are disabled by ourCheck');
  flags.set(kTGnameF);
  print('Set D radio. F will clear; then B and C are enabled by ourCheck');
  flags.set(kTGnameD);
}

const kTGname0 = 0;
const kTGnameA = 1;
const kTGnameB = 2;
const kTGnameC = 3;
const kTGnameD = 4;
const kTGnameE = 5;
const kTGnameF = 6;
const kTGnameG = 7;
const kTGnameH = 8;
const kTGnameI = 9;
const kTGnameJ = 10;
const kTGnameK = 11;
const kTGnameL = 12;
const kTGnameM = 13;
const kTGnameN = 14;
const kTGnameO = 15;
const kTGnameP = 16;
const kTGnameQ = 17;
const kTGnameR = 18;
const kTGnameS = 19;
const kTGnameT = 20;
const kTGnameU = 21;
const kTGnameW = 22;
const kTGnameV = 23;
const kTGnameX = 24;
const kTGnameY = 25;
const kTGnameZ = 26;


example/toggler_example.dart
// Copyright (c) 2022 Wojciech S. Czarnecki, aka Ohir Ripe

import 'package:toggler/toggler.dart';

void main() {
  void ourNotify(_, Toggler cu) {
    var tg = '    set: ';
    var ds = ' active: ';
    for (int i = 0; i < 27; i++) {
      tg += cu[i] ? ' ^' : ' -';
      ds += cu.active(i) ? ' .' : ' !';
    }
    print('kTGname:  0 A B C D E F G H I J K L M N O P Q R S T U W V X Y Z');
    print(tg);
    print(ds);
    print('          -----------------------------------------------------');
  }

  // pt is a copy of previous state, cu is the current (live) state
  bool ourCheck(Toggler oS, Toggler nS) {
    // validate: 'NameA' can toggle only if 'Name0' was set.
    if (!oS[kTGname0] && oS[kTGnameA] != nS[kTGnameA]) {
      print('      >>> NameA change supressed by validator');
      return false; // disallow change to 'NameA'
    }
    // fix B,C if F radio was toggled
    if (oS[kTGnameF] != nS[kTGnameF]) {
      if (nS[kTGnameF]) {
        nS.disable(kTGnameB);
        nS.disable(kTGnameC);
        nS.clear(kTGnameB);
        nS.clear(kTGnameC);
      } else {
        nS.enable(kTGnameB);
        nS.enable(kTGnameC);
      }
    }
    return true; // accept changes
  }

  final flags = Toggler(notify: ourNotify, fix: ourCheck);

  // declare a radioGroup, up to 17 groups can be made over 51 items.
  flags.radioGroup(kTGnameD, kTGnameF);
  // fiddle:
  print('Trying to set A (ourCheck validator disallows this)');
  flags.set(kTGnameA);
  print('Set 0, NameA can be set only if Name0 was set before');
  flags.set(kTGname0);
  print('Now A is allowed to be toggled');
  flags.set(kTGnameA);
  print('Set B');
  flags.set(kTGnameB);
  print('Set C');
  flags.set(kTGnameC);
  print('Set D (of D..F radio group)');
  flags.set(kTGnameD);
  print('Set E of radio D..F - D will clear automatically.');
  flags.set(kTGnameE);
  print('Set F radio. E will clear then B and C are disabled by ourCheck');
  flags.set(kTGnameF);
  print('Set D radio. F will clear; then B and C are enabled by ourCheck');
  flags.set(kTGnameD);
}

/// toMask int extensiom takes index and returns const in with 1 set at index
extension TogglerMask on int {
  toMask(int i) => i >= 0 && i <= tgIndexMax ? 1 << i : 0;
}

/// for use in Rx settings state methods can be added as an extension
extension TogglerRx on Toggler {
  /// apply externally mutated state to the _Model_ object.
  bool apply(Toggler src, {bool doNotify = true, bool force = false}) {
    if (!force && hh > src.hh) return false;
    Toggler? oldS;
    if (doNotify && notify != null) oldS = state();
    tg = src.tg;
    ds = src.ds;
    rg = src.rg;
    hh = src.hh;
    if (doNotify && notify != null) notify!(oldS!, this);
    return true;
  }

  /// unconditionally remove handlers from cloned object
  void freeze() => fix = notify = null;

  /// _compact action byte_ of the most recent change coming from a state setter.
  /// For use with `replay` method below.
  ///
  /// CAbyte keeps _incoming_ changes, not ones made internally by `fix`.
  /// CAbyte layout: `(0/1) b7:tg/ds b6:clear/set b5..b0 change index`
  int get cabyte => hh.toUnsigned(8);

  /// takes compact action byte and applies it - emulating a setter run.
  /// Used for testing and debugging (eg. with actions saved at end user devices).
  void replay(int cas) {
    final i = cas.toUnsigned(6);
    final isDs = cas & 1 << 7 != 0;
    final actS = cas & 1 << 6 != 0;
    final va = isDs
        ? actS
            ? ds |= (1 << i)
            : ds &= ~(1 << i)
        : actS
            ? tg |= (1 << i)
            : tg &= ~(1 << i);
    pump(i, va, isDs, actS);
  }
}

/// always use symbolic index
const kTGname0 = 0;
const kTGnameA = 1;
const kTGnameB = 2;
const kTGnameC = 3;
const kTGnameD = 4;
const kTGnameE = 5;
const kTGnameF = 6;
const kTGnameG = 7;
const kTGnameH = 8;
const kTGnameI = 9;
const kTGnameJ = 10;
const kTGnameK = 11;
const kTGnameL = 12;
const kTGnameM = 13;
const kTGnameN = 14;
const kTGnameO = 15;
const kTGnameP = 16;
const kTGnameQ = 17;
const kTGnameR = 18;
const kTGnameS = 19;
const kTGnameT = 20;
const kTGnameU = 21;
const kTGnameW = 22;
const kTGnameV = 23;
const kTGnameX = 24;
const kTGnameY = 25;
const kTGnameZ = 26;
